scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/BucketMetadata.java

```diff
@@ -403,11 +403,9 @@ private <MetadataT extends BucketMetadata> boolean isIntraPartitionCompatibleWit
     final Class<? extends BucketMetadata> otherClass = other.getClass();
     final Set<Class<? extends BucketMetadata>> compatibleTypes = compatibleMetadataTypes();
 
-    if (compatibleTypes.isEmpty() && other.getClass() != this.getClass()) {
-      return false;
-    } else if (this.getKeyClass() != other.getKeyClass()
```

Comment:
this check was not correct; it should have been comparing the metadata class itself.